packages/xod-project/src/typeDeduction.js

```diff
@@ -0,0 +1,252 @@
+  getOutwardPinKeyFromLink,
+  getOwnPinKeyFromLink,
+  previouslyDeducedTypes,
+  getPatchByNodeId,
+  linksToNode
+) =>
+  R.compose(
+    catMaybies,
+    R.map(link => {
+      const outwardNodeId = getOutwardNodeIdFromLink(link);
+      const outwardPinKey = getOutwardPinKeyFromLink(link);
+
+      // :: Maybe (PinKey, DataType)
+      return R.compose(
+        R.map(R.pair(getOwnPinKeyFromLink(link))),
+        R.chain(
+          R.compose(
+            maybeGetTypeFromPreviouslyDeduced(
+              outwardNodeId,
+              outwardPinKey,
+              previouslyDeducedTypes
+            ),
+            Pin.getPinType
+          )
+        ),
+        Patch.getPinByKey(outwardPinKey),
+        getPatchByNodeId
+      )(outwardNodeId);
+    })
+  )(linksToNode);
+
+const getPinKeysByGenericType = patch =>
+  R.compose(
+    R.map(R.map(Pin.getPinKey)),
+    R.groupBy(Pin.getPinType),
+    R.filter(Pin.isGenericPin),
+    Patch.listPins
+  )(patch);
+
+// returns Map NodeId (Map PinKey (Maybe DataType))
+const deducePinTypesForNode = (
+  abstractNodeId,
+  getPatchByNodeId,
+  previouslyDeducedPinTypes,
+  pinTypesFromLinks
+) => {
+  const abstractPatch = getPatchByNodeId(abstractNodeId);
+  // :: Map DataType [PinKey]
+  const pinKeysByGenericType = getPinKeysByGenericType(abstractPatch);
+
+  // :: Map PinKey (Maybe DataType)
+  const previouslyDeducedPinTypesForNode = R.propOr(
+    {},
+    abstractNodeId,
+    previouslyDeducedPinTypes
+  );
+
+  return R.compose(
+    R.reduce(
+      R.mergeWith(
+        // Nothing means there are are contradictions
+        (a, b) => (a.equals(b) ? a : Maybe.Nothing())
+      ),
+      {}
+    ),
+    R.append(previouslyDeducedPinTypesForNode),
+    R.map(R.map(Maybe.of)),
+    // convert from [(PinKey, DataType)] to [Map PinKey DataType]
+    // and propagate detected types to all pins with the same generic type
+    catMaybies,
+    R.map(([pinKey, dataType]) =>
+      R.compose(
+        R.map(
+          R.compose(
+            R.ifElse(
+              Utils.isGenericType, // TODO: filter out connections to generic pins earlier?
+              // for example, if pins A and B are t1, and we discovered
+              // that A is a number, mark B as a number as well
+              R.compose(
+                R.fromPairs,
+                R.map(pk => [pk, dataType]),
+                R.propOr([], R.__, pinKeysByGenericType)
+              ),
+              R.always(R.objOf(pinKey, dataType))
+            ),
+            Pin.getPinType
+          )
+        ),
+        Patch.getPinByKey(pinKey)
+      )(abstractPatch)
+    )
+  )(pinTypesFromLinks);
+};
+
+// filters out intermediate "broken" input(or output) pins,
+// which we needed only during resolving process
+// :: Map NodeId (Map PinKey (Maybe DataType)) -> Map NodeId (Map PinKey (Maybe DataType))
+const omitIntermediateBrokenPins = (getPatchByNodeId, listPins) =>
```

Comment:
I’m a bit confused about the “broken” term. It’s too wide. Do you mean “Ambiguous”?